mongo-store.js

```diff
@@ -181,40 +185,163 @@ module.exports = function (opts) {
     },
 
     save: function (args, cb) {
-      var ent = args.ent
+      const is_update = Boolean(args.ent.id)
 
-      var update = !!ent.id
+      if (is_update) {
+        return updateExisting(args, cb)
+      }
 
-      getcoll(args, ent, function (err, coll) {
-        if (error(args, err, cb)) return
+      return createAndSave(args, cb)
 
-        var entp = ent.data$(false)
 
-        if (!update) {
-          var id
-          if (undefined !== ent.id$) {
-            id = ent.id$
-          } else if (opts.generate_id) {
-            id = opts.generate_id(ent)
+      function createAndSave(args, cb) {
+        return upsertIfRequested(args, function (err, upsert) {
+          if (error(args, err, cb)) {
+            return
           }
 
-          entp._id = makeid(id)
+          if (upsert.upsert_requested) {
+            return cb(null, upsert.out)
+          }
+
+          return createNew(args, cb)
+        })
+
+
+        function upsertIfRequested(args, cb) {
+          const query_for_save = args.q
+
+          if (Array.isArray(query_for_save.upsert$)) {
+            const upsert_on = query_for_save.upsert$
+            const public_entdata = args.ent.data$(false)
+
+            if (upsert_on.length > 0 && upsert_on.every(p => p in public_entdata)) {
```

Comment:
why the need to check if the upsert prop is in the ent data?
if this fails it's a no-op and there's no feedback
if we pass through "bad" upsert props, mongo should complain, which is what we want

Reply:
@rjrodger Could you please elaborate? Currently, the code needs to check for that so that if it fails, it knows there can be no successful match, and hence no upsert, - in which case it reverts to creating a new entity.

Reply:
Here's a test which serves as an example of the expected behavior:
```javascript
describe('fields in upsert$ are not present in the data$ object', () => {
  beforeEach(() => new Promise(fin => {
    si.make('products')
      .data$({ label: 'a toothbrush', price: '3.40' })
      .save$(fin)
  }))

  it('creates a new entity because it can never match', fin => {
    si.test(fin)

    si.make('products')
      .data$({ label: 'a toothbrush', price: '2.95' })
      .save$({ upsert$: ['label', 'coolness_factor'] }, err => {
        if (err) {
          return fin(err)
        }

        si.make('products').list$({}, (err, products) => {
          if (err) {
            return fin(err)
          }

          expect(products.length).to.equal(2)

          expect(products[0]).to.contain({
            label: 'a toothbrush',
            price: '3.40'
          })

          expect(products[1]).to.contain({
            label: 'a toothbrush',
            price: '2.95'
          })

          return fin()
        })
      })
  })
})
```